### PR TITLE
S2graph 7 abstract common code

### PR DESCRIPTION
**Pull request by daewon**
Abstract common codes for rest project into s2core

Please refer to [S2GRAPH-7](https://issues.apache.org/jira/browse/S2GRAPH-7?jql=project%20%3D%20S2GRAPH%20AND%20priority%20in%20%28Critical%2C%20Major%2C%20Minor%2C%20Trivial%29%20AND%20resolution%20%3D%20Unresolved%20ORDER%20BY%20key%20DESC)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -35,10 +39,8 @@ lazy val s2counter_core = project.dependsOn(s2core)
...
   .settings(commonSettings: _*)
 
-libraryDependencies ++= Seq(
-  ws,
-  filters,
-  "xalan" % "serializer" % "2.7.2", // Download in Intelli J(Download Source/Document)
-  "com.github.danielwegener" % "logback-kafka-appender" % "0.0.3",
-  "org.json4s" %% "json4s-native" % "3.2.11" % Test
-)
+lazy val s2ml = project.settings(commonSettings: _*)
```
Comment (SteamShon) [inline build.sbt]: There is no sub project named s2ml on master branch. can you check this?

File: build.sbt [hunk 2]
```diff
@@ -14,13 +14,17 @@ lazy val commonSettings = Seq(
     "Cloudera" at "https://repository.cloudera.com/artifactory/cloudera-repos",
     "Twitter Maven" at "http://maven.twttr.com",
     "sonatype-snapshots" at "https://oss.sonatype.org/content/repositories/snapshots",
-    "Scalaz Bintray Repo" at "http://dl.bintray.com/scalaz/releases"
+    "Scalaz Bintray Repo" at "http://dl.bintray.com/scalaz/releases",
+    "sonatype-releases" at "https://oss.sonatype.org/content/repositories/releases/"
   )
 )
 
-lazy val root = project.in(file(".")).enablePlugins(PlayScala)
```
Comment (SteamShon) [inline build.sbt]: so there is no root project. previous root has been changed into s2rest_play right?

File: project/plugins.sbt [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
-
 // Use the Play sbt plugin for Play projects
 addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.3.10")
 
 // http://www.scalastyle.org/sbt.html
 addSbtPlugin("org.scalastyle" %% "scalastyle-sbt-plugin" % "0.7.0")
 
-resolvers += "sonatype-releases" at "https://oss.sonatype.org/content/repositories/releases/"
+// sbt revolver
+addSbtPlugin("io.spray" % "sbt-revolver" % "0.8.0")
```
Comment (SteamShon) [inline project/plugins.sbt:8]: Why do we need this?

File: s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala [hunk 4]
```diff
@@ -309,13 +312,14 @@ object Graph {
   } get
 }
 
-class Graph(_config: Config)(implicit ec: ExecutionContext) {
+class Graph(_config: Config)(implicit val ec: ExecutionContext) {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala]: Why execution context of Graph is exposed into member?

File: s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala [hunk 5]
```diff
@@ -309,13 +312,14 @@ object Graph {
...
   val config = _config.withFallback(Graph.DefaultConfig)
   val cacheSize = config.getInt("cache.max.size")
 //  val cache = CacheBuilder.newBuilder().maximumSize(cacheSize).build[java.lang.Integer, Seq[QueryResult]]()
   val vertexCache = CacheBuilder.newBuilder().maximumSize(cacheSize).build[java.lang.Integer, Option[Vertex]]()
 
-  Model(config)
+  Model.apply(config)
+  Model.loadCache()
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala:322]: :+1:

File: s2core/src/main/scala/com/kakao/s2graph/core/PostProcess.scala [hunk 6]
```diff
@@ -38,23 +45,23 @@ object PostProcess extends JSONParser {
       edgesWithRanks = edgesAndRanks.groupBy(x => x._2.srcVertex).map(_._2.head)
       id <- innerValToJsValue(target, tgtColumn.columnType)
     } yield {
-        Json.obj("name" -> tgtColumn.columnName, "id" -> id,
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/PostProcess.scala]: This is just reformat right?

File: s2core/src/main/scala/com/kakao/s2graph/core/PostProcess.scala [hunk 7]
```diff
@@ -298,10 +305,10 @@ object PostProcess extends JSONParser {
         parentQueryParam = QueryParam(parentEdge.labelWithDir)
         parents = edgeParent(parentEdge.parentEdges, q, parentQueryParam) if parents != JsNull
       } yield {
-          val originalEdge = parentEdge.originalEdgeOpt.getOrElse(parentEdge)
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/PostProcess.scala]: This is also just reformat?

File: s2core/src/main/scala/com/kakao/s2graph/core/mysqls/Model.scala [hunk 8]
```diff
@@ -54,6 +54,19 @@ object Model {
       res
     }
   }
+
+  def shutdown() = {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/mysqls/Model.scala:58]: :+1:

File: s2core/src/test/scala/com/kakao/s2graph/core/Integrate/QueryTest.scala [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+                "direction": "out",
+                "offset": $offset,
+                "limit": $limit
+              }
+            ]]
+          }
+          """)
+
+  // called by each test, each
+  override def beforeEach = initTestData()
```
Comment (SteamShon) [inline s2core/src/test/scala/com/kakao/s2graph/core/Integrate/QueryTest.scala:501]: :+1:

File: s2rest_play/build.sbt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+name := "s2rest_play"
+
+version := "0.12.1-SNAPSHOT"
```
Comment (SteamShon) [inline s2rest_play/build.sbt]: Can we use Common.Version instead of hard coded version?

File: build.sbt (hunk 1 above)
Comment (daewon) [inline build.sbt]: @SteamShon My mistake, i'll remove it

File: project/plugins.sbt (hunk 3 above)
Comment (daewon) [inline project/plugins.sbt:8]: @SteamShon 
sbt-revolver for automatically restart application if sources have been changed

[sbt-revolver](https://github.com/spray/sbt-revolver)

File: s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala (hunk 4 above)
Comment (daewon) [inline s2core/src/main/scala/com/kakao/s2graph/core/Graph.scala]: @SteamShon I thought it necessary for test code. but now it's not necessary, i'll remove

**Comment by daewon** [comment]
@SteamShon i did apply your review.

File: build.sbt (hunk 1 above)
Comment (hsleep) [inline build.sbt]: @daewon Maybe you missed above comment.

File: project/plugins.sbt (hunk 3 above)
Comment (SteamShon) [inline project/plugins.sbt:8]: @daewon I understand. it totally makes sense.

**Comment by SteamShon** [comment]
@daewon looks good to me.

**Comment by daewon** [comment]
This branch has been Merged to https://github.com/apache/incubator-s2graph/pulls

**Pull request closed without merging**